### DISCUSSION TITLE
`consult-line` vs buffer-local completion styles

### DISCUSSION THREAD

**Discussion by jdtsmith** [minad/vertico#533] [General]
I specialize `orderless-style-dispatcher` locally in some buffers (for corfu).  The values I set it to are appropriate for in-buffer completion only.  This normally works fine, since `vertico` in the mini-buffer uses its own (global) value for that variable.  

But when using `vertico-multiform=buffer`, the completion styles are actually run in the original buffer, not in a mini-buffer, so you have different search semantics depending on whether something like `consult-line` is conducted in a buffer, or a mini-buffer (e.g. `grid`).

**Comment by minad** [comment]
> But when using vertico-multiform=buffer, the completion styles are actually run in the original buffer, not in a mini-buffer, so you have different search semantics depending on whether something like consult-line is conducted in a buffer, or a mini-buffer (e.g. grid).

No, I doubt that this is true. vertico-buffer does not change the buffer where the completion is executed. Vertico always runs inside the minibuffer. Vertico wouldn't even work otherwise since its state is stored in minibuffer local variables.

EDIT: Thinking about this again, I understand where your confusion is coming from, and I haven't explained this before. `consult-line` executes the completion style both in the minibuffer and in the original buffer. This is a `consult-line` specialty needed for finding the match location and is not related to Vertico/vertico-buffer/etc. See https://github.com/minad/consult/blob/7a7af8dcdda02b9aa4a680a228b3f3a5cfa95334/consult.el#L3307-L3310. Unfortunately there is no way to avoid this, since `consult-line` has to run after the minibuffer has already been closed. Maybe a potential solution could be to address this problem in `consult-line` only, basically at its root, temporarily copying/let-binding the minibuffer completion style settings in the original buffer when performing the match. But then, `consult-focus-lines` and `consult-keep-lines` also run in the original buffer with the same `consult-location` category, so your in-buffer settings for Corfu are futile anyway.

> I specialize orderless-style-dispatcher locally in some buffers (for corfu). The values I set it to are appropriate for in-buffer completion only. This normally works fine, since vertico in the mini-buffer uses its own (global) value for that variable.

Unfortunately such setups are unsupported. They simply don't fit in the existing completion style machinery. One solution is to set a completion style globally, then set completion styles locally inside the Corfu buffers and finally override the `consult-location` completion category again with the global completion style. While this solution works it is neither elegant nor looks like a proper solution which we hand over to users as the "proper way" to do it.

Whenever I think about better solutions I come up with empty hands, just now again. I think the issue ideally has low priority, which is the case if you configure your completion styles and categories well. For example Orderless works well enough just everywhere for me. This hadn't been the case always, and I had to apply certain optimizations and improvements over the years to Orderless. See for example https://github.com/oantolin/orderless/issues/96 and https://github.com/oantolin/orderless/pull/179.

**Reply by minad** [reply]
See https://github.com/minad/consult/pull/1131 which tries to address the problem in Consult. This way one can configure completion styles globally for minibuffer completion and buffer locally for in-buffer Corfu completion. I believe that this should work, except if the user configures completion styles buffer locally inside the minibuffer. But this setting had never been supported with `consult-line` anyway. Do you see any problems with this approach?

**Reply by jdtsmith** [reply]
I was confused about this; it does happen when transferring `consult-line` to the mini-buffer as well.  I think using the global settings makes good sense.  The only problem I'd see is if some users have configured completion styles (or other variables, like `orderless-style-dispatcher`) locally, and expect that to apply for `consult-line`, etc., thinking of those as "buffer-local searches", compared to minibuffer searches like `consult-buffer`.  Could be an option.

**Reply by minad** [reply]
No, buffer-local consult-line settings have never been possible since consult-line requires the buffer and minibuffer settings to be consistent.

**Reply by minad** [reply]
I've merged the Consult patch and added some documentation to Corfu: https://github.com/minad/corfu/commit/53a5f116254500fe1883a4179b0154f219ef59b2

Btw, in case you have time to read over the Corfu documentation and improve it, that would be very welcome. In particular since I am not a native English speaker.
